rfcs/0000-kuadrant_crd.md

```diff
@@ -0,0 +1,164 @@
+currently enabled. `UPDATE` and `DELETE` should remain unhindered, as in order to transition the
+policy support requires no policies to be present.
+
+If any of the following steps fails, all previously applied changes need to be unrolled and the
+`status` block of the CR would clearly reflected what changed and that Kuadrant is currently _not_
+operational. This is to be considered an atomic operation, either all gets successfully deployed or
+nothing is.
+
+If a policy is disabled but a policy is being added, when disallowing the admission, the webhook would
+use a response code `501 Not Implemented` with the additional `message` indicating the policy is
+disabled and as such can't be honored.
+
+### In an OSSM cluster
+
+The deployment of `Kuadrant` should fail if `authentication` isn't explicitly disabled.
+
+## State transitions of policy support
+
+If the `Kuadrant` object gets updated to change the support of a policy, the change would only be
+considered if no such policy exists in the cluster. No state transitions with regards to policy
+support is possible if policies of the given type are present.
+
+While it would be good to be able to enable the support for a policy while such policies already
+exist, there is much to consider as to what this would entail - here are a few things to consider:
+
+- can we apply all these policies?
+- what ordering should be considered to apply them?
+- what would a conflict mean?
+- what would disabling `authentication` for a cluster, while existing routes are currently protected by these?
+
+Requiring to first remove all policies and then only safely disable or enable the support for that
+kind of policy seems reasonable at this stage and can always be added later, which clearly specified
+behavior.
+
+### Disabling the support for a policy
+
+When disabling the support for a policy, merely the admission webhook's behavior changes. The
+`Authorino` resource or `Limitador` ones remain untouched.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The flipside of the flexibility this provides users with, this introduces quite some complexity. Hence
+the desire to keep this very rigid as to what are the preconditions to change the support for a
+policy.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Possibly the gain of not running unnecessary services on the cluster isn't big enough for us to
+justify adding this complexity. Yet, as of today, we need a way to disable `authentication` in the
+case of running within an OSSM backed cluster.
+
+We could disable `authentication` on a heuristic base approach: running within OSSM? Disable
+`authentication`. The user experience seems less than desirable and confusing as best, as in future
+version this would probably change. That being said we're still pre-v1. We'd also need to find a way
+to know very early on what environment we are being deployed in.
```

Comment:
Hmm. So this for me comes to the fundamental question. Do we want to support these different system states as feature or do we want to say in v0.x.y authpolicy is not supported in a particular environment add the webhook mentioned earlier during installation and reject any AuthPolicies in that environment. 
It also raises an interesting question for me that if we want to offer individual features that perhaps RateLimitPolicy and AuthPolicy should become a part of those components (IE Limitador/Limitador operator and Authorino/Authorino operator) should support AuthPolicy and RateLimitPolicy respectively. Have we given that approach any consideration? This would mean enabling in the kuadrant operator would install or uninstall the operator and in turn would install or remove the CRD, if memory serves removing the CRD will remove any of those resource types also. Doing this way would mean that we would not need the webhook as the CRD is either their or not.
cc @eguzki @guicassolato

Reply:
I would not follow the approach of adding/removing CRDs dynamically based on some Kuadrant CR. CRD's are global cluster objects and that raises potential conflicts. What if some user in the cluster is using the Limitador operator to limit their service by their own (not using kuadrant)? Or how does this idea fit when kuadrant supports multiple kuadrant data plane instances?

Regarding having AuthPolicy part of the Authorino Operator (or RateLimitPolicy part of the Limitador Operator) is interesting to consider. I do not see now a hard stopper.

Reply:
I don't see Kuadrant's RLP and KAP as definitions of Limitador Operator and Authorino Operator respectively. These CRDs of Kuadrant are committed to GW-API's policy attachment value proposition, which the aforementioned operators aren't. In fact, I'd see such as a violation of the current dependency that exists between Kuadrant and its functional components, where the lower dependencies (Limitador and Authorino) currently can exist without knowing anything about Kuadrant or its purpose.

Furthermore and especially within the scope of this particular RFC, I agree with @eguzki's point of CRDs being global definitions in the cluster and therefore their installation not the way to control whether a particular workload in a particular namespace shall be deployed or not.

For both of these reasons, I believe that _installing_ the CRD is one thing, while choosing which functional component to _deploy_ is another separate one. Therefore, RLP and KAP CRDs should continue living in the Kuadrant Operator IMO.